Add support for custom headers and migrate from legacy auth

Jeetulsamaiya opened the pull request:
This PR solves this issue:
 #756 #750 #549 #249 

- Introduced CustomHeaders component for managing custom headers.
- Updated Sidebar to include CustomHeaders component.
- Enhanced useConnection hook to handle custom headers and legacy auth migration.
- Added tests for custom headers functionality.
- Updated package.json and package-lock.json for new dependencies.

<!-- Provide a brief summary of your changes -->

## Motivation and Context
<!-- Why is this change needed? What problem does it solve? -->
This change addresses the need for flexible authentication and header management in the Inspector UI. Previously, the application only supported a single bearer token with a custom header name. This enhancement allows users to configure multiple custom headers with key-value pairs, providing better support for various authentication schemes and API requirements. The implementation also maintains backward compatibility by automatically migrating existing legacy auth configurations to the new custom headers system.

## How Has This Been Tested?
<!-- Have you tested this in a real application? Which scenarios were tested? -->
- Added comprehensive unit tests for the custom headers functionality in `useConnection.test.tsx`
- Tested migration from legacy bearer token authentication to custom headers
- Verified that disabled headers are properly ignored
- Tested both form-based and JSON-based header configuration modes
- Ensured backward compatibility with existing authentication configurations
- Tested the UI components for adding, editing, and removing custom headers

## Breaking Changes
<!-- Will users need to update their code or configurations? -->
No breaking changes. The implementation maintains full backward compatibility with existing bearer token and header name configurations. Legacy auth settings are automatically migrated to the new custom headers format when the application loads.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->
The implementation includes:
- A new `CustomHeaders` component with both form-based and JSON editor modes
- Type definitions for custom headers in `customHeaders.ts`
- Migration utilities for backward compatibility with legacy auth
- Enhanced `useConnection` hook to process multiple headers
- Server-side support for multiple custom auth headers via `x-custom-auth-headers`
- UI improvements with toggle switches for enabling/disabling individual headers
- Password-style input fields with visibility toggles for sensitive header values

The feature supports both streamable HTTP and SSE transports, allowing users to configure authentication headers for various API endpoints and authentication schemes beyond simple bearer tokens.


Fixes #756
Fixes #750
Fixes #549 
Fixes #249

Jeetul-samaiya committed:
Add support for custom headers and migrate from legacy auth

- Introduced CustomHeaders component for managing custom headers.
- Updated Sidebar to include CustomHeaders component.
- Enhanced useConnection hook to handle custom headers and legacy auth migration.
- Added tests for custom headers functionality.
- Updated package.json and package-lock.json for new dependencies.

Jeetulsamaiya commented:
<img width="1920" height="1080" alt="Screenshot from 2025-08-27 00-22-11" src="https://github.com/user-attachments/assets/08bea2ea-42ce-4202-8707-97a93bd20c75" />
<img width="1920" height="1080" alt="Screenshot from 2025-08-27 00-22-08" src="https://github.com/user-attachments/assets/39ef472d-f508-4265-89d7-ef73542a2ff9" />
<img width="1920" height="1080" alt="Screenshot from 2025-08-27 00-21-51" src="https://github.com/user-attachments/assets/0de26096-37df-4c84-8d69-3c28214ca070" />

these are the design screenshots

alex210501 commented:
I think it would makes sense to also add those configuration field to the `mcp.json` file too. I'm not strongly familiar with the code, so maybe you already did it and I just did not see it.

Jeetul-samaiya committed:
refactor: improve code formatting and readability across multiple files

Jeetul-samaiya committed:
fix: update error handling in JSON parsing and add ESLint rule for unused variables

Jeetul-samaiya committed:
fix: standardize formatting in bug report issue template

olaservo reviewed (commented):
Hi, thank you for this PR!  I just had a minor comment so far.  I wanted to test this locally but don't have a server ready at the moment that requires custom auth headers, so just wanted to leave a note that I'm not ignoring this one while I set up something to test with locally (unless you have a specific server you have been testing this with?)

Jeetul-samaiya committed:
fix: enhance error handling for custom headers and improve test coverage for header management

ryanchase-cohere commented:
Hi @olaservo 

Thank you for your support on this! A simple way to test this would be to host an MCP servers on Google Cloud Run adn make them private because Cloud Run expects a user's google identity token to be sent as the `X-serverless-authorization` header, while the MCP server itself in most cases relies on the `Authorization` header for custom authorization.  Example: https://cloud.google.com/run/docs/authenticating/service-to-service#acquire-token

ryanchase-cohere commented:
> Hi @olaservo
> 
> Thank you for your support on this! A simple way to test this would be to host an MCP server on Google Cloud Run and make it private because Cloud Run expects a user's google identity token to be sent as the `X-Serverless-Authorization` header, while the MCP server itself in most cases relies on the `Authorization` header for custom authorization. https://cloud.google.com/run/docs/authenticating/service-to-service#acquire-token

here's an example: https://cloud.google.com/blog/topics/developers-practitioners/build-and-deploy-a-remote-mcp-server-to-google-cloud-run-in-under-10-minutes

Jeetulsamaiya committed:
Merge branch 'main' into main

Jeetulsamaiya committed:
Merge branch 'main' into main

Jeetul-samaiya committed:
feat: add default Authorization Bearer header and clean up legacy auth

- Provide "Authorization: Bearer " as default starting header
- Remove legacy auth props from components
- Centralize migration logic and fix localStorage cleanup
- Update tests for new auth architecture

cliffhall reviewed (approved):
Requested changes were made. Tested locally, observed request headers in browser and at the server endpoint.
 LGTM! 👍

## Request headers in browser
<img width="310" height="71" alt="Screenshot 2025-09-15 at 11 52 17 AM" src="https://github.com/user-attachments/assets/05eaea69-7173-4439-b619-d1eac2559a28" />

## Request headers at server
<img width="642" height="997" alt="Screenshot 2025-09-15 at 12 10 00 PM" src="https://github.com/user-attachments/assets/8b64ab32-2e58-4d1e-9889-c77acd53ce5e" />

cliffhall commented:
@olaservo You don't necessarily need the server to expect the headers, as long as you can see that they are verifiably sent and received by the server. I did that [in my review](https://github.com/modelcontextprotocol/inspector/pull/751#pullrequestreview-3225352175) and everything looks good.
